File: Tree/428-LC-199-rightSideView.py
# Definition for a binary tree node.
from collections import deque


class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
    def addLeft(self,node):
        self.left=node
    def addRight(self,node):
        self.right=node
        
class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if not root:
            return root
        q=deque([root])
        ans=[]
        while q:
            ans.append(q[-1].val)
            n=len(q)
            for _ in range(n):
                x=q.popleft()
                if x.left:
                    q.append(x.left)
                if x.right:
                    q.append(x.right)
                
        return ans


nodes=[1,2,3,4,5]
root=TreeNode(1)
l1=TreeNode(2)
r1=TreeNode(3)
l2=TreeNode(4)
r2=TreeNode(5)
root.addLeft(l1)
root.addRight(r1)
l1.addLeft(l2)
l1.addRight(r2)
print(Solution().rightSideView(root))
    

File: Tree/428-LC-199-rightSideView.cpp
#include<iostream>
#include<vector>
#include<queue>
using VEC=std::vector<int>;

class TreeNode
{
    private:
        int m_val{};
        TreeNode *m_left{nullptr};
        TreeNode *m_right{nullptr};
    public:
        TreeNode(int val,TreeNode* left=nullptr, TreeNode* right=nullptr):m_val{val},m_left{left},m_right{right}{}
        void addLeft(TreeNode* left){m_left=left;}
        void addRight(TreeNode* right){m_right=right;}
        TreeNode* getLeft() {return m_left;}
        TreeNode* getRight() {return m_right;}
        int getVal() const {return m_val;}
};

VEC rightSide(TreeNode* root)
{
    VEC ans{};
    if(!root)
        return ans;

    std::queue<TreeNode*> q{};
    q.push(root);
    while(!q.empty())
    {
        ans.push_back(q.back()->getVal());
        int n=q.size();
        for(int i=0;i<n;++i)
        {
            TreeNode* curNode=q.front();
            q.pop();

            if(curNode->getLeft())
                q.push(curNode->getLeft());
            if(curNode->getRight())
                q.push(curNode->getRight());

        }
    }
    return ans;
}

int main()
{
    VEC nums{1,2,3,4,5};
    TreeNode *root=new TreeNode(1);
    TreeNode *l1=new TreeNode(2);
    TreeNode *r1=new TreeNode(3);
    TreeNode *l2=new TreeNode(4);
    TreeNode *r2=new TreeNode(5);

    root->addLeft(l1);
    root->addRight(r1);
    l1->addLeft(l2);
    l1->addRight(r2);
    VEC res=rightSide(root);
    for(auto &v:res)
        std::cout<<v<<',';
}

